shell/server/backup.js

```diff
@@ -41,6 +41,33 @@ function recursiveRmdirIfExists(dir) {
   }
 }
 
+function walkAndWriteFiles(dir, writeStream, replaceRoot, originalRoot) {
+  originalRoot = originalRoot || dir;
+  var files = Fs.readdirSync(dir);
+  files.forEach(function (name) {
+    name = dir + "/" + name;
+
+    // Filter out file paths with a newline in them
+    if (name.indexOf("\n") === -1) {
+      return;
+    }
+
+    var fileStat = Fs.lstatSync(name);
+    if (fileStat.isDirectory()) {
+      walkAndWriteFiles(name, writeStream, replaceRoot, originalRoot);
+    } else if (fileStat.isFile()) {
+      name = replaceRoot + name.slice(originalRoot.length);
+      writeStream.write(name + "\n");
+    }
+  });
+
+  // Include empty directories
+  if (files.length === 0) {
+    dir = replaceRoot + dir.slice(originalRoot.length);
```

Comment:
I think it's fine. Any call to walkAndWriteFiles except for the first can only occur within the forEach loop, after the check for the new line has already happened. This should ensure `dir` can never have a newline in it, except for on the first call. We enter walkAndWriteFiles with a known path (specifically Path.join(SANDSTORM_GRAINDIR, grainId, "sandbox")), so it should be fine.

That being said, the logic is a bit twisty and I'm not against adding another check just for clarity.